### ISSUE TITLE
Possible 

### ISSUE THREAD

**Issue by felipebalbi** [protesilaos/denote#54]

**Comment by felipebalbi** [comment]
For reference, `denote--id-format` and `denote--id-regexp` would change to the following values:

```lisp
(defconst denote--id-format "%Y%m%dT%H%M%S%N"
  "Format of ID prefix of a note's filename.")

(defconst denote--id-regexp "\\([0-9]\\{8\\}\\)\\(T[0-9]\\{15\\}\\)"
  "Regular expression to match `denote--id-format'.")
```

**Comment by protesilaos** [comment]
Hello @felipebalbi!

You are right: in this scenario we do have a conflict.  Though I wonder what the use-case is.  Is it realistic to expect someone to create multiple notes per second?  How does this work in practice?

> This can be solved by append `%N` to `denote--id-format`, would that be an acceptable change? It would probably require a "migration" utility to automatically rename all existing notes to append the nanoseconds field.

Yes, this is what would be required.

**Comment by felipebalbi** [comment]
> Hello @felipebalbi!

Hi @protesilaos 

> You are right: in this scenario we do have a conflict.  Though I wonder what the use-case is.  Is it realistic to expect someone to create multiple notes per second?  How does this work in practice?

Right, it's very academic in nature. The possibility exists, but one needs to desire to reproduce it. I decided to report it because solving it is trivial, however it has a direct impact on existing notes. You may wish to ignore this as a "non-bug".

**Comment by protesilaos** [comment]
> Right, it's very academic in nature. The possibility exists, but one needs to desire to reproduce it. I decided to report it because solving it is trivial, however it has a direct impact on existing notes. You may wish to ignore this as a "non-bug".

You did well to report it.  It is good to discuss what we have and be aware of what the possible problems are.

When I was deciding on what to include in the date/identifier, I thought going up to seconds was a good compromise betwen precision and practicality.

However, this scenario can come up when renaming files.  It is an edge-case, but it is possible.  So we need to consider how best to approach it if/when it happens.  Perhaps we could add `%N` only when a conflict arises?  That way we do not burden day-to-day workflows with the extra numbers?

I say we keep this issue open in case someone else wants to speak their mind.

**Comment by jeanphilippegg** [comment]
Just a comment to add an alternative solution. I did not try to create a collision but, if it happens, we could increment the seconds by 1 until there is no collision. A human user should not notice.

**Comment by protesilaos** [comment]
> we could increment the seconds by 1 until there is no collision. A human user should not notice.

Indeed.  It would also mean that all names look the same.  Maybe that would become a burden at scale though as we would keep rolling the problem over to the next second?  I have in mind the scenario where, say, 100 notes are created.  This is why I was wondering about the use-case.

At any rate, I like the idea.

Another option is to produce an error as we already do with `denote--barf-duplicate-id`.  Obviously, this does not help the use-case, but it ensures we do not have duplicates.

**Comment by protesilaos** [comment]
Just a sec.  When I try this:

```elisp
(defun test ()
  (denote "Test")
  (denote "Test"))

(test)
```

I get one file with duplicate front matter.  I thought we would have two files.

**Comment by jeanphilippegg** [comment]
Since it is within the same second, it is normal. We should check in `denote--prepare-note` that the file we want to add a front matter to does not already exist.

**Comment by protesilaos** [comment]
The reason we do not have two files is because `denote--prepare-note` uses `find-file`, though it's late for me and I am getting a bit fuzzy now.

Conceptually, I agree that we do not handle duplicates past the second.  Though now I am curious to actually produce the problem, regardless of how we may choose to deal with it.

**Comment by protesilaos** [comment]
> We should check in `denote--prepare-note` that the file we want to add a front matter to does not already exist.

Yes, we could.  Just to be safe.

**Comment by jeanphilippegg** [comment]
I would like to refactor `denote--prepare-note` a little. There is a check for the `path` parameter in 3 places and we don't use it in `denote`. This is not related, but since we modify the same function, I would start with this. Then, I will check this current issue. Is it good with you?

**Comment by protesilaos** [comment]
> Is it good with you?

Yes.  Thank you! In general, all refinements are welcome.  No need to ask.

**Comment by jeanphilippegg** [comment]
> Just a sec. When I try this:
> 
> ```emacs-lisp
> (defun test ()
>   (denote "Test")
>   (denote "Test"))
> 
> (test)
> ```
> 
> I get one file with duplicate front matter. I thought we would have two files.

There is a function `denote--barf-duplicate-id` that should have thrown a user error. I believe the reason it is not working is because the file are not save at the end of the `denote` command. So it cannot find the file. This is something we will have to think about.

**Comment by felipebalbi** [comment]
> > Just a sec. When I try this:
> > ```emacs-lisp
> > (defun test ()
> >   (denote "Test")
> >   (denote "Test"))
> > 
> > (test)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get one file with duplicate front matter. I thought we would have two files.
> 
> There is a function `denote--barf-duplicate-id` that should have thrown a user error. I believe the reason it is not working is because the file are not saved at the end of the `denote` command. So it cannot find the file. This is something we will have to think about.

perhaps check for the existence of a file or a buffer with the file's name?

**Comment by protesilaos** [comment]
Hello again folks.  I am closing this issue.  If there ever is an actual bug where Denote creates duplicate identifiers, please report it.
